# Author : Jonathan Rondeau-Leclaire
# This new version of tax_glom updates the index names to the 
# agglomerated taxrank, to avoid e.g. keeping an arbitrary ASV for
# a genus that encompasses multiple ASVs post agglomeration. 

# Finally, in the event that some species have missing taxrank,
# those are simply removed entirely.

### Further detail : 
# I created what I think is a better version of this function, 
# which I creatively called tax_glom2(). To explain why, consider three things:

# 1. tax_glom() will always keep the  original OTU names (taxa_names) regardless
# of the agglomeration. This means, if you have 20 OTU in the same Genus, you 
# end up with one Genus (expected), but its taxa_name will be one of the OTUs 
# (probably the first one the algorithm encounters). With shotgun data that's 
# especially annoying because the phyloseq OTU value (the taxa_name attribute)
# will be a species. These are the unique keys that link your otu_table to your 
# tax_table. Then, when you use software that uses these (e.g. for differential 
# abundance analysis), they output the name of a species when, in fact, you have
# a genus-agglomerated phyloseq object. So my first enhancement is for 
# tax_glom2() to generate a phyloseq object with the taxa_names updated to 
# reflect the taxrank you agglomerate to.

# 2. with more and more MAGs being added to public databases, there are 
# instances where an organism will have a species-level label, but no genus or 
# family. An example :  Clostridia bacterium UC5.1-1E11 only has a class-level 
# classification; thus Order, Family, and Genus will all be NA. So keeping my 
# previous modification to tax_glom in mind, agglomeration at the genus level 
# will pool ALL read counts of NA genera together, regardless of their higher 
# taxonomy. Since these genomes have not been classified at these levels (all 
# one can say is that it's a "species" of the class Clostridia, probably because 
# marker genes from that clade were found in that genome; and that it is different 
# enough from all other known clostridia genomes so that we know it's an unknown 
# species of Clostridia. In this case, if you are interested in genus-level 
# metagenomics, this genome is as good as an undetected one, so tax_glom2() 
# simply removes these counts when you agglomerate to Genus level.

# 3. Agglomerating creates a new taxa which encompasses all the counts 
# of all taxa that are NA or 'Unclassified' ; therefore, they are
# removed post-agglomeration and counts go down.

# Comments mostly and shamelessly generated by ChatGPT 4o
# Function to melt, aggregate, and rebuild a phyloseq object
tax_glom2 <- function(ps, taxrank) {
  require('phyloseq')
  require('tidyverse')
  # Ensure the phyloseq object has a taxonomy table
  if (is.null(tax_table(ps))) {
    stop("The function requires a taxonomy table.")
  }
  
  # Ensure the provided taxrank exists
  if (!taxrank %in% rank_names(ps)) {
    stop("Invalid taxrank. Must be one of rank_names(ps).")
  }
  
  # Melt the OTU table to long format
  otu_long <- otu_table(ps) %>% 
    data.frame(check.names = FALSE) %>% # avoid sample name modification
    rownames_to_column('Sample') %>% 
    tidyr::pivot_longer(-Sample, names_to = "OTU", values_to = "Abundance")
  
  # Extract taxonomy and join it with the OTU data
  tax_long <- tax_table(ps) %>% data.frame %>% 
    dplyr::filter(!is.na(!!sym(taxrank))) %>% 
    rownames_to_column('OTU')
  
  # Extract all taxranks up to taxrank
  taxrank_idx <- match(taxrank, colnames(tax_long))

  # Join OTU and taxonomy data; left_join ensures that taxa
  # with missing taxrank are removed
  merged_long <- left_join(tax_long[,1:taxrank_idx], 
                           otu_long, 
                           by = "OTU")
  
  # Aggregate counts by Sample and taxrank (sum)
  agg_data <- merged_long %>%
    group_by(Sample, !!sym(taxrank)) %>%
    dplyr::summarise(Abundance = sum(Abundance, na.rm = TRUE), 
              .groups = "drop") %>% 
    filter(!!sym(taxrank) != 'Unclassified' & !is.na(!!sym(taxrank)))
  
  # Pivot the data back to wide format (Samples as columns, taxa as rows)
  otu_matrix <- agg_data %>%
    tidyr::pivot_wider(names_from = "Sample", 
                values_from = "Abundance", 
                values_fill = 0) %>%
    arrange({{ taxrank }}) %>% 
    column_to_rownames(var = taxrank) %>%
    as.matrix()
  
  # Create a new taxonomy table with only the selected taxrank
  tax_matrix <- tax_long[,1:taxrank_idx] %>% 
    dplyr::select(-OTU) %>%
    as.data.frame %>% 
    mutate(tax1=!!sym(taxrank)) %>% 
    distinct(!!sym(taxrank), .keep_all = TRUE) %>%  ## Some weird instances of same genera from different families...
    arrange(tax1) %>% 
    column_to_rownames('tax1') %>% 
    as.matrix
  
  # Rebuild the phyloseq object with the aggregated OTU and taxonomy tables
  new_ps <- phyloseq(
    otu_table(otu_matrix, taxa_are_rows = TRUE),
    tax_table(tax_matrix),
    sample_data(ps@sam_data)
  ) %>% 
  prune_samples(sample_sums(.) > 0, .) %>%  #remove any empty samples 
    prune_taxa(taxa_sums(.) > 0,.) #remove taxa absent from all
  
  return(new_ps)
}
